dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/AnnotatedJar.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.opensource.classpath;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.nio.file.Path;
+import java.util.Objects;
+import javax.annotation.Nullable;
+import org.eclipse.aether.artifact.Artifact;
+
+/** JAR file annotated with an optional Maven artifact. */
+final class AnnotatedJar {
```

Comment:
What do you think about renaming this `CachedArtifact` or `LocalArtifact` instead? That is, it's an artifact that has a JAR rather than a JAR that has coordinates?

Though now that I look deeper, `org.eclipse.aether.artifact.Artifact` has a `File` property that's serving the same purpose as the path here. Can we just use that? Then we don't need to add this class at all.

Reply:
The idea of _not_ using `org.eclipse.aether.artifact.Artifact` is to support Linkage Checker invocation without Maven artifacts. That is `LinkageCheckerMain --jar A.jar,B.jar`.

Reply:
In that case we could use an Artifact with group ID, artifact ID, etc. st to null:

https://maven.apache.org/resolver/apidocs/org/eclipse/aether/artifact/DefaultArtifact.html

Reply:
That would technically work. However, I don't think it's good idea to use Artifact class to represent something that is not a Maven artifact.

Reply:
I suppose it's more of a generalized thing that includes the class of Maven and non-Maven artifacts. Or it's a special kind of artifact that has a file and null coordinates. It does feel like this class is redundant.

Reply:
@elharo What do you think about dropping bare JAR support? I've implemented that but did not encounter any need to perform operations for bare JAR files; when I use Linkage Checker, it's always through Maven artifacts.

Reply:
I checked [DefaultArtifact constructor that takes a file](https://maven.apache.org/resolver/apidocs/org/eclipse/aether/artifact/DefaultArtifact.html#DefaultArtifact(java.lang.String,%20java.lang.String,%20java.lang.String,%20java.lang.String,%20java.lang.String,%20java.util.Map,%20java.io.File)). 6 out of 7 arguments would null if we only store a file there. And it's `java.io.File`, not `java.nio.file.Path`.  I added an example usages of AnnotatedJar and DefaultArtifact in [a comment](https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/1303#issuecomment-601869672). 

With the reason above, I prefer this AnnotatedJar to DefaultArtifact.

Reply:
Think of it this way. This is inside out and redundant. All the information we need is in Artifact. If you add a third field for Dependency or DependencyNode, then all the information is in there. But this type doesn't add any new information. It's not like we're adding things that aren't in the artifact like download time or something like that. Does that make sense?

Reply:
That's a great explanation. Now I feel the redundancy of AnnotatedJar clearly when we use Linkage Checker for Maven artifacts.

How about cases without Maven artifacts (bare JAR files)? Do you have rationale for using Maven's Artifact class for JAR files that do not carry `groupId`, `artifactId`, or `version`?

If we had algebraic data types, I could just write:

```
data ClassPathElement = Path | Artifact  
```